threescale/src/utils.rs

```diff
@@ -1,13 +1,58 @@
-use crate::structs::Application;
+use crate::structs::{ Application, PeriodWindow };
+use proxy_wasm::hostcalls::{ set_shared_data, get_shared_data};
+use log::info;
+use std::time::{ SystemTime, Duration };
 
 // Returns Application from shared data with CAS integer
 pub fn get_application_from_cache(key: &str) -> Option<(Application,u32)> {
-    None
+    match get_shared_data(&key).unwrap()
+    {
+        (bytes,cas) => {
+            match bytes {
+                Some(data) => Some((bincode::deserialize::<Application>(&data).unwrap(),cas.unwrap())),
```

Comment:
Note that it is possible that `cas` might be `None` for this case, and `unwrap()` would then panic. If you prefer to deal with `cas` elsewhere, you might want to change the signature to return `Option<(Application, Option<u32>)>`.

Reply:
Fixed.